xds/src/main/java/io/grpc/xds/PriorityLoadBalancer.java

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.CONNECTING;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.READY;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+import static io.grpc.xds.XdsSubchannelPickers.BUFFER_PICKER;
+
+import io.grpc.ConnectivityState;
+import io.grpc.InternalLogId;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.internal.ServiceConfigUtil.PolicySelection;
+import io.grpc.util.ForwardingLoadBalancerHelper;
+import io.grpc.util.GracefulSwitchLoadBalancer;
+import io.grpc.xds.PriorityLoadBalancerProvider.PriorityLbConfig;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/** Load balancer for priority policy. */
+final class PriorityLoadBalancer extends LoadBalancer {
+  private final Helper helper;
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService executor;
+  private final XdsLogger logger;
+
+  // Includes all active and deactivated children. Mutable. New entries are only added from priority
+  // 0 up to the selected priority. Entries are only deleted when priority lb is shutdown or 15
+  // minutes after the entry's deactivation.
+  private final Map<String, ChildLbState> children = new HashMap<>();
+
+  // Following fields are only null initially.
+  private ResolvedAddresses resolvedAddresses;
```

Comment:
It's better to cache addresses and mappings in a way that is easy for this class to use and represents this class's internal state, instead of dumping `ResolvedAddresses` here.

Reply:
We still need `resolvedAddresses.toBuilder()` for `ChildLbState.updateResolvedAddresses()`. Otherwise we need cache all fields, such as `attributes`,  of `resolvedAddresses` that are unchanged and passing downstream. If `ResolvedAddresses` have more fields in the future, it's tedious to cache them and easy forgotten.

Reply:
Personally, I am not wild of doing `resolvedAddresses.toBuilder()` for passing addresses to downstream policies. I would just create a new `ResolvedAddresses` and only put in things needed. 

Each policy should keep only information it and its downstream policies need from `ResolvedAddresses` and throw away anything else. It should only pass down information needed by its downstream policies, nothing more. We are doing address filtering to give each downstream policy only addresses they need, same idea should apply for all information they would receive. This is how layering works.

But anyway, we already have a lot of this kind of "lazy ideas" involved in the design, I am fine with what it is now.

Reply:
Each policy does not know in general what information are not needed by its downstream. Each policy can only process the information it knows (address filtering, and exactly child policy), all other unrecognized attributes need be pass downstream.